src/cryptonote_core/service_node_swarm.cpp

```diff
@@ -0,0 +1,260 @@
+#include "service_node_swarm.h"
+
+#undef LOKI_DEFAULT_LOG_CATEGORY
+#define LOKI_DEFAULT_LOG_CATEGORY "service_nodes"
+
+#ifdef UNIT_TEST
+  #define prod_static
+#else
+  #define prod_static static
+#endif
+
+namespace service_nodes
+{
+  static uint64_t get_new_swarm_id(std::mt19937_64 &mt, const swarm_snode_map_t &swarm_to_snodes)
+  {
+    uint64_t id_new;
+    do
+    {
+      id_new = uniform_distribution_portable(mt, UINT64_MAX);
+    } while (swarm_to_snodes.count(id_new) != 0);
+
+    return id_new;
+  }
+
+  /// The excess is calculated as the total number of snodes above MIN_SWARM_SIZE across all swarms
+  prod_static size_t calc_excess(const swarm_snode_map_t &swarm_to_snodes)
+  {
+    const size_t excess = std::accumulate(swarm_to_snodes.begin(),
+                                          swarm_to_snodes.end(),
+                                          size_t(0),
+                                          [](size_t result, const swarm_snode_map_t::value_type &pair) {
+                                            const ssize_t margin = pair.second.size() - EXCESS_BASE;
+                                            return result + std::max(margin, ssize_t(0));
+                                          });
+    LOG_PRINT_L2("Calculated excess: " << excess);
+    return excess;
+  };
+
+  /// Calculate threshold above which the excess should create a new swarm.
+  /// The threshold should be such that
+  /// 1. there is enough excess to create a new swarm of size NEW_SWARM_SIZE AND
+  /// 2. there is enough excess to leave IDEAL_SWARM_MARGIN excess in the existing swarms
+  prod_static size_t calc_threshold(const swarm_snode_map_t &swarm_to_snodes)
+  {
+    const size_t threshold = NEW_SWARM_SIZE + (swarm_to_snodes.size() * IDEAL_SWARM_MARGIN);
+    LOG_PRINT_L2("Calculated threshold: " << threshold);
+    return threshold;
+  };
+
+  prod_static excess_pool_snode pick_from_excess_pool(const std::vector<excess_pool_snode>& excess_pool, std::mt19937_64 &mt)
+  {
+    /// Select random snode
+    const auto idx = uniform_distribution_portable(mt, excess_pool.size());
+    return excess_pool.at(idx);
```

Comment:
Are you sure it's not `[0, size[` ?

From what I could understand:
```cpp
 do x = mersenne_twister(); while (x >= secureMax);
return  x / (secureMax / n);
```
So first, `x` is always smaller than `secureMax`, so `x / secureMax < 1`.

Now, because:
```cpp
secureMax = mersenne_twister.max() - mersenne_twister.max() % n;
```
`secureMax` will always be a multiple of n, thus `secureMax / n` will always be a nice clean division.
That way, `x / (secureMax / n) == (x / secureMax * n) < (1 * n)`.

Online example:
http://cpp.sh/4hjao

Reply:
I also think it is [0, n).

Reply:
I just demonstrated it's not lol

Reply:
I'm not sure what you mean. `[0, n)` means including 0 and excluding n.

Reply:
I can contradict this by using N as 18446744073709551615, i.e. mersenne_twister::max or `uint64_max`

So

```
mersenne_twister::max() = uint64_max
n = uint64_max

secureMax = uint64_max - (uint64_max % uint64_max) = uint64_max
do ...
return x / (secureMax / uint64_max) => return (x / 1) => return x
```

Oh and er, my comment was incorrect, I meant what Maxim said, [0, n)

Reply:
Waiit. It returns x, which is >= secureMax. Nevermind.

Reply:
Ok I think we all agree haha.
My notation `[0, size[` means "between a, inclusive, and b exclusive". I didn't know the `[0, size)` notation, sorry, so I assumed it was a typo for `[0, size]`!
EDIT: According to wikipedia, I am the one with the weird notation...
` The notation `[a, b)` is used [...]. In some European countries, the notation `[5,12[` is also used for this`